Added first solution to the problem

k1eu opened the pull request:
Due to an issue with API , other (Countries) API has been used.

Known problems.
- Tests are simply bad.. and not working
- Next Button isn't disabled on last page

k1eu committed:
Added first solution to the problem
Due to an issue with API , other (Countries) API has been used

k1eu committed:
fix: added missing files

Wozniaxos commented on src/App.js:
```diff
@@ -1,24 +1,19 @@
+    <Router>
+      <Switch>
+        <Route path="/countries" component={Countries}/>
+        <Route path="/country/:countryId" component={Country} />
```
In a proper semantic way it should be `/countries/:id`

k1eu replied:
Changed to `/countries/:id` in incoming commit

Wozniaxos commented on src/components/CountriesList.js:
```diff
@@ -0,0 +1,27 @@
+
+  return (
+    <div>
+      {data.countries.map((country, index) => (
```
If index is not used we can remove it

k1eu replied:
Fixed in incoming commit!

Wozniaxos commented on src/components/CountriesList.js:
```diff
@@ -0,0 +1,27 @@
+          </Link>
+          <h3>Languages used:</h3>
+          <ul>
+            {country.languages.map((language, index) => (
```
Same as above

k1eu replied:
Fixed in incoming commit!

Wozniaxos commented on src/components/Country.test.js:
```diff
@@ -0,0 +1,98 @@
+    request: {
+      query: COUNTRY_QUERY,
+      variables: {
+        Id: "Q889",
```
I think that lower case, just simple `id` might be better here, in case of longer words, camelCase as a rule of thumbs :)

k1eu replied:
Changed to lowercase `id` in incoming commit ;)

Wozniaxos commented on src/pages/country.js:
```diff
@@ -0,0 +1,54 @@
+
+  if (loading) return <div>Loading...</div>
+  if (error) return <div>something is wrong</div>
+  console.log('dupa') // ;-)
```
I like that

k1eu replied:
Glad you liked it, I made it just for you! ;)

Wozniaxos commented on src/index.css:
```diff
@@ -1,5 +1,6 @@
 body {
   margin: 0;
+  padding-left: 15px;
```
15? let's give it 16 :D

k1eu replied:
I swapped it for 1rem. Is it any better?

Wozniaxos commented on src/index.js:
```diff
@@ -1,14 +1,24 @@
 
-ReactDOM.render(
-  <React.StrictMode>
+const API_KEY = '25f50ab3-2426-447f-a4dc-a521a96ec52e'
```
In real world it should be set at env level to be able to change it on production environtment but in this scenario this is cool 👍

Wozniaxos commented on src/pages/countries.js:
```diff
@@ -0,0 +1,82 @@
+      }
+    }
+  }
+`;
```
In real world we will probably keep queries encapsulated in files but this is totally ok in that scenarion 👍

k1eu replied:
I encapsulated the queries into different files in incoming commit

Wozniaxos commented on src/pages/countries.js:
```diff
@@ -0,0 +1,82 @@
+
+function useQueryUrl() {
+  return new URLSearchParams(useLocation().search);
+}
```
In real scenario it can be more convenient, for example above function may return object with params that you can later destructure, so the usage might be `const { page } = useQueryUrl()` but this is also cool

Wozniaxos commented on src/pages/countries.js:
```diff
@@ -0,0 +1,82 @@
+  const { loading, error, data } = useQuery(COUNTRIES_QUERY, variables);
+
+  const getNextPageNumber = () => {
+    const currPage = parseInt(query.get("page")) || 0;
```
i think that this variable can be. extracted with `useMemo(() => {})` hook

k1eu replied:
I changed the currPage into a variable extracted from useMemo hook in incoming commit

Wozniaxos commented on src/pages/countries.js:
```diff
@@ -0,0 +1,82 @@
+    return nextPage;
+  };
+  const getPreviousPageNumber = () => {
+    const currPage = parseInt(query.get("page"));
```
and used also here

Wozniaxos commented on src/pages/countries.js:
```diff
@@ -0,0 +1,82 @@
+      const previousPage = currPage - 1;
+      return previousPage;
+    } else {
+      return 0;
```
I think that this validation logic should be in handleClick function because in the current situation we can do `history.push` a lot of time with `0` page, better to prevent history pushing at all in that case

k1eu replied:
I remodeled functions  for handling next and previous page. Now I only have `handleClick` functions without  `getPreviousPageNumber` etc.

Wozniaxos commented on src/pages/countries.js:
```diff
@@ -0,0 +1,82 @@
+        <button onClick={() => handleOnClickPrevious()} disabled={page <= 0}>
+          Wróć
+        </button>
+        <button onClick={() => handleOnClickNext()} disabled={page > 20}>
```
I guess its only for demonstration purposes since we don't know how many pages it has in the `skip` scenario. I asked cause the logic of disabling next after reaching the last page number is a little bit weird, let's say that you have 20 pages, so in your logic it will disable it at 21 page which may not exist, but I guess it's the only demonstration disable so no worries 👍

Wozniaxos replied:
Same with handler as above

Wozniaxos commented on src/pages/countries.js:
```diff
@@ -0,0 +1,82 @@
+    <div>
+      <p>Cinema page</p>
+      <div>
+        <button onClick={() => handleOnClickPrevious()} disabled={page <= 0}>
```
For this and the last page we can make 2 variables with `useMemo` like `onFirstPage` and `onLastPage` to be more verbose :)

Wozniaxos replied:
And the most important, you don't need to do `{() => handleOnClickPrevious()}` you can just do `{handleOnClickPrevious}` it is bad idea to put anonymous function in handlers since every rerender this function is created in the memory which may cause memory leaks

Wozniaxos commented on src/pages/country.js:
```diff
@@ -0,0 +1,54 @@
+  return (
+    <div>
+      {data ?
+        (data.countries.map((country, index) => (
```
We don't need index

Wozniaxos replied:
I am wondering if in the case that we are sure that this is a  `one` country cause we are looking it through `id` maybe a better option would be to do `countries[0]` instead of mapping which suggests the collection? It can be misleading during reading of the code

k1eu replied:
Yeah, it's much better. I swapped mapping through array with one object, for having data directly from `countries[0]`

Wozniaxos commented on src/pages/country.js:
```diff
@@ -0,0 +1,54 @@
+            )}
+            <h2>The cities in given country:</h2>
+            <ul>
+              {country.cities.map((city, index) => (
```
No need of index

Wozniaxos reviewed (commented):
Great job overall! Some small improvement comments :)

k1eu committed:
refactor: many changes that follow the pr comments

ecapsulated queries into separate folder
view showing country details is not on /countries/:id path
changed padding to rems
refactored functions on countries page
countries page variables are now from useMemo hook
buttons onClick functions are now called directly instead of being called by anonymous function
country page doesnt map through array of one object, instead it gets data from the first object in array
deletion of many index variables from map functions that are not being utilized
